### ISSUE TITLE
Using threads with webpack in node application

### ISSUE THREAD

**Issue by chingyawhao** [andywer/threads.js#222]
I'm getting this error when spawning a worker in my application

```
Error: Cannot find module 'internal/process/worker.ts'
Require stack:
- /Users/chingyawhao/Documents/project/node_modules/threads/dist/master/implementation.node.js
- /Users/chingyawhao/Documents/project/node_modules/threads/dist/master/implementation.js
- /Users/chingyawhao/Documents/project/node_modules/threads/dist/master/index.js
- /Users/chingyawhao/Documents/project/node_modules/threads/dist/index.js
- /Users/chingyawhao/Documents/project/dist/main.js
- /Users/chingyawhao/Documents/project/node_modules/node-hot-loader/HmrServer.js
- /Users/chingyawhao/Documents/project/node_modules/node-hot-loader/loader.js
- /Users/chingyawhao/Documents/project/node_modules/node-hot-loader/node-hot.js
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:982:15)
    at Function.resolve (internal/modules/cjs/helpers.js:83:19)
    at resolveScriptPath (/Users/chingyawhao/Documents/project/node_modules/threads/dist/master/implementation.node.js:63:19)
    at new Worker (/Users/chingyawhao/Documents/project/node_modules/threads/dist/master/implementation.node.js:74:40)
    at eval (webpack:///./src/command/music/index.ts?:43:57)
    at Generator.next (<anonymous>)
    at eval (webpack:///./src/command/music/index.ts?:8:71)
    at new Promise (<anonymous>)
    at __awaiter (webpack:///./src/command/music/index.ts?:4:12)
    at runMusicPlayer (webpack:///./src/command/music/index.ts?:40:42)
    at exports.default (webpack:///./src/command/music/index.ts?:51:13)
    at executeCommand (webpack:///./src/bronya.ts?:36:13)
    at eval (webpack:///./src/bronya.ts?:46:17)
    at processTicksAndRejections (internal/process/task_queues.js:97:5) {
  code: 'MODULE_NOT_FOUND'
```

This is how I spawn the worker
```
const runMusicPlayer = async() => {
    console.log('attempting')
    try {
        const hello = await spawn(new Worker('./worker.ts', {type:'module'}))
        const data = await hello('Hello')
        console.log(data)
    } catch(error) {
        console.error(error)
    }
}
runMusicPlayer()
```

I tried `await spawn(new Worker(require.resolve('./worker.ts'), {type:'module'}))` too
Doesn't seem to work
`Error: Cannot find module 'internal/process/worker.ts'` now becomes `Error: Cannot find module 'internal/process/src/command/worker.ts'`

**Comment by andywer** [comment]
Hey @chingyawhao. Can you please state the node.js version you are using?

**Comment by chingyawhao** [comment]
My node.js version is v12.16.1

**Comment by chingyawhao** [comment]
@andywer I think I found the potential problem
I added `"module": "esnext"` to my tsconfig and I'm now getting a different error message

```
Error: Module build failed (from ./node_modules/threads-plugin/dist/loader.js):
ModuleNotFoundError: Module not found: Error: Can't resolve 'child_process' in '/Users/chingyawhao/Documents/project/node_modules/tiny-worker/lib'
    at /Users/chingyawhao/Documents/project/node_modules/webpack/lib/Compilation.js:925:10
    at /Users/chingyawhao/Documents/project/node_modules/webpack/lib/NormalModuleFactory.js:401:22
    at /Users/chingyawhao/Documents/project/node_modules/webpack/lib/NormalModuleFactory.js:130:21
    at /Users/chingyawhao/Documents/project/node_modules/webpack/lib/NormalModuleFactory.js:224:22
    at /Users/chingyawhao/Documents/project/node_modules/neo-async/async.js:2830:7
    at /Users/chingyawhao/Documents/project/node_modules/neo-async/async.js:6877:13
    at /Users/chingyawhao/Documents/project/node_modules/webpack/lib/NormalModuleFactory.js:214:25
    at /Users/chingyawhao/Documents/project/node_modules/webpack/node_modules/enhanced-resolve/lib/Resolver.js:213:14
    at /Users/chingyawhao/Documents/project/node_modules/webpack/node_modules/enhanced-resolve/lib/Resolver.js:285:5
    at eval (eval at create (/Users/chingyawhao/Documents/project/node_modules/webpack/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:13:1)
    at eval (webpack:///./src/command/music/music-player.ts?./node_modules/threads-plugin/dist/loader.js?%7B%22name%22:%220%22%7D:1:7)
    at Object.call (/Users/chingyawhao/Documents/project/dist/main.js:857:1)
    at __webpack_require__ (/Users/chingyawhao/Documents/project/dist/main.js:691:30)
    at fn (/Users/chingyawhao/Documents/project/dist/main.js:60:20)
    at eval (webpack:///./src/command/music/index.ts?:84:41)
    at Module.call (/Users/chingyawhao/Documents/project/dist/main.js:1107:1)
    at __webpack_require__ (/Users/chingyawhao/Documents/project/dist/main.js:691:30)
    at fn (/Users/chingyawhao/Documents/project/dist/main.js:60:20)
    at eval (webpack:///./src/bronya.ts?:10:72)
    at Module.call (/Users/chingyawhao/Documents/project/dist/main.js:881:1)
    at __webpack_require__ (/Users/chingyawhao/Documents/project/dist/main.js:691:30)
    at fn (/Users/chingyawhao/Documents/project/dist/main.js:60:20)
    at eval (webpack:///./src/index.ts?:5:65)
    at Module.call (/Users/chingyawhao/Documents/project/dist/main.js:1691:1)
    at __webpack_require__ (/Users/chingyawhao/Documents/project/dist/main.js:691:30)
    at fn (/Users/chingyawhao/Documents/project/dist/main.js:60:20)
```

It seems like webpack is trying to resolve child_process
I already have webpack-node-externals and placed it in externals of webpack.config.js

**Comment by andywer** [comment]
Interesting… Can you please share the version of threads-plugin that you use? It should be >= 1.3.0 as there were some bug fixes recently.

**Comment by chingyawhao** [comment]
Yup...it is
```
"threads": "^1.3.1",
"tiny-worker": "^2.3.0",
"threads-plugin": "^1.3.1",
```

**Comment by andywer** [comment]
I think I have an idea what to try. It's again webpack and the child compiler behaving differently out-of-the-box than the root compiler – looks like it doesn't set the `externals` to the common built-in node modules for the child compiler, even if the `target` is `node`.

Can you try the following? Monkey-patch your `node_modules/threads-plugin/lib/loader.js` and change the following line of code [here](https://github.com/andywer/threads-plugin/blob/30789fd3cbfedcd3cb2aa7dc34413dc848379f02/src/loader.js#L61):

```diff
- if (pluginOptions.target === "electron-node-worker") {
+ if (/node/.test(pluginOptions.target)) {
```

**Comment by chingyawhao** [comment]
@andywer I just tried replacing the line on both dist and src folder
Doesn't seems to work
The error message is exactly the same

**Comment by andywer** [comment]
Just to make sure: Have you set your webpack `target` to `node`?

You can also add a `console.log()` to make sure the line with the `NodeTargetPlugin` is run.

**Comment by chingyawhao** [comment]
I tried adding this to my webpack.config
```
new ThreadsPlugin({
    target: 'electron-node-worker'
})
```

Now it yields a different error message
```
Error: Cannot find module 'internal/process/0.worker.js'
Require stack:
- /Users/chingyawhao/Documents/project/node_modules/threads/dist/master/implementation.node.js
- /Users/chingyawhao/Documents/project/node_modules/threads/dist/master/implementation.js
- /Users/chingyawhao/Documents/project/node_modules/threads/dist/master/index.js
- /Users/chingyawhao/Documents/project/node_modules/threads/dist/index.js
- /Users/chingyawhao/Documents/project/dist/main.js
- /Users/chingyawhao/Documents/project/node_modules/node-hot-loader/HmrServer.js
- /Users/chingyawhao/Documents/project/node_modules/node-hot-loader/loader.js
- /Users/chingyawhao/Documents/project/node_modules/node-hot-loader/node-hot.js
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:982:15)
    at Function.resolve (internal/modules/cjs/helpers.js:83:19)
    at resolveScriptPath (/Users/chingyawhao/Documents/project/node_modules/threads/dist/master/implementation.node.js:63:19)
    at new Worker (/Users/chingyawhao/Documents/project/node_modules/threads/dist/master/implementation.node.js:74:40)
    at eval (webpack:///./src/index.ts?:44:95)
    at Generator.next (<anonymous>)
    at eval (webpack:///./src/index.ts?:12:71)
    at new Promise (<anonymous>)
    at __awaiter (webpack:///./src/index.ts?:8:12)
    at runMusicPlayer (webpack:///./src/index.ts?:41:42)
    at __webpack_exports__.default (webpack:///./src/index.ts?:53:13)
    at executeCommand (webpack:///./src/index.ts?:37:13)
    at eval (webpack:///./src/index.ts?:47:17)
    at processTicksAndRejections (internal/process/task_queues.js:97:5) {
  code: 'MODULE_NOT_FOUND'
```

**Comment by chingyawhao** [comment]
1. My webpack config target is indeed set to 'node'
2. Where should I put the console.log in? `node_modules/threads-plugin/lib/loader.js`?

**Comment by andywer** [comment]
1. Nevermind, I think I was wrong there – you already did the right thing, I think :)
2. Yes, please.

**Comment by chingyawhao** [comment]
@andywer the line with `NodeTargetPlugin` is running

**Comment by chingyawhao** [comment]
Hi @andywer 
I'm trying out NodeTemplatePlugin to replace WebWorkerTemplatePlugin
```if (/node/.test(pluginOptions.target) || pluginOptions.target === "electron-node-worker") {
  new NodeTargetPlugin().apply(workerCompiler);
  new NodeTemplatePlugin(workerOptions).apply(workerCompiler);
} else {
  new WebWorkerTemplatePlugin(workerOptions).apply(workerCompiler);
}```

I'm still getting the same error on runtime
```Error: Cannot find module 'internal/process/0.worker.js'
Require stack:
- /Users/chingyawhao/Documents/project/node_modules/threads/dist/master/implementation.node.js
- /Users/chingyawhao/Documents/project/node_modules/threads/dist/master/implementation.js
- /Users/chingyawhao/Documents/project/node_modules/threads/dist/master/index.js
- /Users/chingyawhao/Documents/project/node_modules/threads/dist/index.js
- /Users/chingyawhao/Documents/project/dist/main.js
- /Users/chingyawhao/Documents/project/node_modules/node-hot-loader/HmrServer.js
- /Users/chingyawhao/Documents/project/node_modules/node-hot-loader/loader.js
- /Users/chingyawhao/Documents/project/node_modules/node-hot-loader/node-hot.js
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:982:15)
    at Function.resolve (internal/modules/cjs/helpers.js:83:19)
    at resolveScriptPath (/Users/chingyawhao/Documents/project/node_modules/threads/dist/master/implementation.node.js:63:19)
    at new Worker (/Users/chingyawhao/Documents/project/node_modules/threads/dist/master/implementation.node.js:74:40)
    at eval (webpack:///./src/index.ts?:44:95)
    at Generator.next (<anonymous>)
    at eval (webpack:///./src/index.ts?:12:71)
    at new Promise (<anonymous>)
    at __awaiter (webpack:///./src/index.ts?:8:12)
    at runMusicPlayer (webpack:///./src/index.ts?:41:42)
    at __webpack_exports__.default (webpack:///./src/index.ts?:53:13)
    at executeCommand (webpack:///./src/index.ts?:37:13)
    at eval (webpack:///./src/index.ts?:47:17)
    at processTicksAndRejections (internal/process/task_queues.js:97:5) {
  code: 'MODULE_NOT_FOUND'```

Why would it look for module 'internal/process/0.worker.js' and not '0.worker.js'
File 0.worker.js is generated

**Comment by andywer** [comment]
> Why would it look for module 'internal/process/0.worker.js' and not '0.worker.js'

@chingyawhao I was asking myself the very same thing… And then it just appeared to me: Because it tries to do something like `require.resolve("0.worker.js")` (Note: Be aware hat it's `"0.worker.js"`, not `"./0.worker.js"`).

Can you please add some logging to `master/implementation.node.js` to print the resolved path that is passed to the `worker_threads` Worker constructor?

**Comment by chingyawhao** [comment]
@andywer I'm adding a `console.log` to `master/implementation.node.js`
```
function resolveScriptPath(scriptPath) {
    console.log('scriptPath: ' + scriptPath)
    // eval() hack is also webpack-related
    const workerFilePath = typeof __non_webpack_require__ === "function"
        ? __non_webpack_require__.resolve(path.join(eval("__dirname"), scriptPath))
        : require.resolve(rebaseScriptPath(scriptPath, /[\/\\]worker_threads[\/\\]/));
    return workerFilePath;
}
```

I'm getting this `scriptPath: 0.worker.js`, so this is what you meant by its trying to resolve the wrong module?

**Comment by chingyawhao** [comment]
Hai @andywer
This function `rebaseScriptPath(scriptPath, /[\/\\]worker_threads[\/\\]/)` is returning `internal/process/0.worker.js` when I log it

**Comment by andywer** [comment]
@chingyawhao  Yep, the issue comes from `rebaseScriptPath`. You can quickly try this:

```diff
- const workerFilePath = typeof __non_webpack_require__ === "function"
-   ? __non_webpack_require__.resolve(path.join(eval("__dirname"), scriptPath))
-   : require.resolve(rebaseScriptPath(scriptPath, /[\/\\]worker_threads[\/\\]/))
+ const relative = filePath => filePath.startsWith(".") || filePath.startsWith("/") ? filePath : "./" + filePath
+ const workerFilePath = typeof __non_webpack_require__ === "function"
+   ? __non_webpack_require__.resolve(relative(path.join(eval("__dirname"), scriptPath)))
+   : require.resolve(relative(rebaseScriptPath(scriptPath, /[\/\\]worker_threads[\/\\]/)))
```

I will try to come up with a patch release by tonight! 🙂

**Comment by chingyawhao** [comment]
Wait @andywer, I think its not working yet
It's just changing `internal/process/0.worker.js` to `./internal/process/0.worker.js`

I think the problem lies with this in `rebaseScriptPath`
```
    const parentCallSite = callsites_1.default().find((callsite) => {
        const filename = callsite.getFileName();
        return Boolean(filename && !filename.match(ignoreRegex) && !filename.match(/[\/\\]master[\/\\]implementation/));
    });
```

**Comment by chingyawhao** [comment]
The value of `parentCallSite` is `internal/process/task_queues.js` which is why `0.worker.js` is now `internal/process/0.worker.js`

Many of the `callsite.getFileName()` is null, like `webpack:///./src/command/music/index.ts?:12:71` which is the correct parent call site I think

**Comment by andywer** [comment]
Hmm. A quick-fix might be:

```diff
  const parentCallSite = getCallsites().find((callsite: CallSite) => {
    const filename = callsite.getFileName()
-   return Boolean(filename && !filename.match(ignoreRegex) && !filename.match(/[\/\\]master[\/\\]implementation/))
+   return Boolean(filename && !filename.match(ignoreRegex) && !filename.match(/[\/\\]master[\/\\]implementation/)) && !filename.match(/^internal\/process/)
  })
```

**Comment by andywer** [comment]
Maybe the issue is also related to this check that kind-of tests whether this code is in a webpack bundle: `typeof __non_webpack_require__ === "function"`.

It seems to be `false` in your case. Maybe replacing this condition by something better (any ideas?) might solve the whole issue at once without any other code changes.

**Comment by chingyawhao** [comment]
@andywer, I'm still having error and I'm starting to get more confused now
With me implementing all your suggested fix, I'm getting `./0.worker.js` but it still fails
`Error: Cannot find module './0.worker.js'`
`Error: Cannot find module '0.worker.js'`

I'm now confused about where the worker module is supposed to be

**Comment by chingyawhao** [comment]
I can see the file `0.worker.js` in my compiled folder

**Comment by andywer** [comment]
If you log `console.log(eval("__dirname"))`, what do you see? Is it the directory that contains the `0.worker.js` file?

**Comment by chingyawhao** [comment]
`eval("__dirname")` is pointing to `/Users/chingyawhao/Documents/project/node_modules/threads/dist/master/`

**Comment by andywer** [comment]
So you use webpack to bundle your code, but `threads` is not included in the bundle…

Then maybe there is nothing wrong with `threads` after all. Is your code still `new Worker(require.resolve('./worker.ts'), {type:'module'})`? If not, try it with the `require.resolve()` in the `new Worker()`.

**Comment by chingyawhao** [comment]
My code was `new Worker('./worker.ts', {type:'module'})`
I try `new Worker(require.resolve('./worker.ts'), {type:'module'})` just now and I'm getting `Error: Cannot find module './src/command/music/worker.ts'`, and I no longer see `0.worker.js` being compiled

**Comment by chingyawhao** [comment]
I use `webpack-node-externals` to exclude webpack from bundling node modules

**Comment by andywer** [comment]
Ohh, sure… The `threads-plugin` only works if the first argument passed to `new Worker()` is a static string.

You can revert it to `new Worker('./worker.ts', {type:'module'})`, though, and try one more time if applying an early `require.resolve()` on the worker path in the constructor solves the issue. If so we can draft a more permanent solution.

Man, I wouldn't have thought that this issue would take so long to fix… 🙈😅

**Comment by chingyawhao** [comment]
Hi @andywer, I don't get what you mean by 
> try one more time if applying an early `require.resolve()` on the worker path in the constructor solves the issue

I don't know where else to put `require.resolve()` besides `new Worker(require.resolve('./worker.ts'), {type:'module'})`

**Comment by andywer** [comment]
Sorry, I meant in the constructor in `implementation.node.js`. I didn't have much time yesterday.

Can you apply all the fixes up to [this point](https://github.com/andywer/threads.js/issues/222#issuecomment-602592370) and then try this in `implementation.node.js`?

```diff
  class Worker extends NativeWorker {
    constructor(scriptPath, options) {
-     const resolvedScriptPath = resolveScriptPath(scriptPath)
+     const resolvedScriptPath = path.join(options. _baseURL, resolveScriptPath(scriptPath))

      if (resolvedScriptPath.match(/\.tsx?$/i) && detectTsNode()) {
        super(createTsNodeModule(resolvedScriptPath), { eval: true })
      } else {
        super(resolvedScriptPath, options)
     }
```

Then do in your code: `new Worker("./worker", { _baseURL: eval("__dirname") })`
Maybe that does the trick…

**Comment by chingyawhao** [comment]
Woohoo...I think it works

Ok, I'll list down the stuff that works
First in `threads-plugin/loader.js`
```
- if (pluginOptions.target === "electron-node-worker") {
+ if (/node/.test(pluginOptions.target) || pluginOptions.target === "electron-node-worker") {
  new NodeTargetPlugin().apply(workerCompiler);
}
```

Then in `threads/master/implementation.node.ts`
```
function rebaseScriptPath(scriptPath, ignoreRegex) {
  const parentCallSite = callsites_1.default().find((callsite) => {
    const filename = callsite.getFileName();
-    return Boolean(filename && !filename.match(ignoreRegex) && !filename.match(/[\/\\]master[\/\\]implementation/));
+    return Boolean(filename && !filename.match(ignoreRegex) && !filename.match(/[\/\\]master[\/\\]implementation/)) && !filename.match(/^internal\/process/);
  });
  const callerPath = parentCallSite ? parentCallSite.getFileName() : null;
  const rebasedScriptPath = callerPath ? path.join(path.dirname(callerPath), scriptPath) : scriptPath;
  return rebasedScriptPath;
}
```

```
-function resolveScriptPath(scriptPath) {
+function resolveScriptPath(scriptPath, options) {
  // eval() hack is also webpack-related
  const workerFilePath = typeof __non_webpack_require__ === "function"
    ? __non_webpack_require__.resolve(path.join(eval("__dirname"), scriptPath))
-    : require.resolve(rebaseScriptPath(scriptPath, /[\/\\]worker_threads[\/\\]/));
+    : require.resolve(path.join(options._baseURL || '.', rebaseScriptPath(scriptPath, /[\/\\]worker_threads[\/\\]/)));
  return workerFilePath;
}
```

```
class Worker extends NativeWorker {
  constructor(scriptPath, options) {
-    const resolvedScriptPath = resolveScriptPath(scriptPath);
+    const resolvedScriptPath = resolveScriptPath(scriptPath, options || {});
    if (resolvedScriptPath.match(/\.tsx?$/i) && detectTsNode()) {
      super(createTsNodeModule(resolvedScriptPath), { eval: true });
    }
```

I think thats all

**Comment by chingyawhao** [comment]
Doing `const resolvedScriptPath = path.join(options. _baseURL, resolveScriptPath(scriptPath))` in the worker constructor doesn't work because the error was thrown inside `resolveScriptPath` when `require.resolve` is finding non existing file

**Comment by andywer** [comment]
Great stuff, @chingyawhao! 🎉

Let's get a PR ready later today. Just need to tweak `resolveScriptPath` a little more, so the same logic applies to the other expression in the ternary and so it still works if `rebaseScriptPath()` returns an absolute path :)

**Comment by andywer** [comment]
Fixed by #224
